src/sentry/api/endpoints/organization_member/index.py

```diff
@@ -137,10 +137,15 @@ def validate_role(self, role):
         return self.validate_orgRole(role)
 
     def validate_orgRole(self, role):
-        if role not in {r.id for r in self.context["allowed_roles"]}:
+        role_obj = next((r for r in self.context["allowed_roles"] if r.id == role), None)
+        if role_obj is None:
             raise serializers.ValidationError(
                 "You do not have permission to set that org-level role"
             )
+        if not self.context.get("allow_retired_roles", True) and role_obj.is_retired:
```

Comment:
Currently `allow_retired_roles` defaults to True but I feel like it will be useful to reuse this option in other places where `OrganizationMemberRequestSerializer` is invoked. For example, this one could be moved to the validator but in a separate PR:
https://github.com/getsentry/sentry/blob/b36f3f9a51c32e0c156c7beaaec207bbb08623f7/src/sentry/api/endpoints/organization_member/details.py#L298-L306